src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -0,0 +1,194 @@
+package com.puppetlabs.puppetserver.pool;
+
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import com.puppetlabs.puppetserver.pool.LockablePool;
+
+/**
+ * A data structure to be used as a Pool, encapsulating a LinkedBlockingDeque
+ * and a ReentrantReadWriteLock. Implements the <tt>LockablePool</tt>
+ * interface such that when a <tt>borrowItem</tt> is performed an instance is
+ * taken from the deque and a read lock is acquired, when a
+ * <tt>returnItem</tt> is performed an instance is put back onto the deque and
+ * a read lock is released, and <tt>lock</tt> locks a write lock.
+ *
+ * @param <E> the type of element that can be added to the LinkedBlockingDeque.
+ */
+public final class JRubyPool<E> implements LockablePool<E> {
+    private final LinkedBlockingDeque<E> liveQueue;
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
+    private final Set<E> registeredElements = new CopyOnWriteArraySet<>();
+
+    public JRubyPool(int size) {
+        liveQueue = new LinkedBlockingDeque<>(size);
+    }
+
+    /**
+    * This method is analagous to <tt>putLast</tt> in the
+    * <tt>LinkedBlockingDeque</tt> class, but also causes the element to be
+    * added to the list of "registered" elements that will be returned by
+    * <tt>getRegisteredInstances</tt>.
+    *
+    * Note that this method is synchronized to try to ensure that the addition
+    * to the queue and the list of registered instances are visible roughly
+    * atomically to consumers, but because the underlying queue uses its own
+    * lock, it is possible for it to be modified on another thread while this
+    * method is being executed.
```

Comment:
That sounds fine (and better).  TBH I don't think it is worth spending a bunch of time on this right now so I'd say change it to that and then let's put this to bed.